### PR TITLE
HTTP request & WebSocket message streaming

### LINKED ISSUE #1764: Overhaul Streaming Support
ref #856
ref #1147
ref #1506

Recent core refactoring efforts touched our streaming support for large responses a couple of times.
Here a couple of proposals which could improve things a lot:
  * support request streaming
     - useful for file uploads, inspecting S3 clients and other HTTP-based storage APIs
     - see #1506 for an unfortunately now obsolete PR
  * stream by default for files larger than X bytes
     - AFAIK the current default is to just fail the request if the returned `Content-Length` is larger than `--body-size-limit`. This is not very convenient, nor intuitive.
  * support WebSocket streaming for messages and/or frames
     - typical WebSocket messages are small JSON-ish things, but one could do large binary file transfers as well.

### PR DESCRIPTION
**Pull request by ujjwal96**
* [x] http1
* [x] http1 tests
* [x] http2
* [x] http2 tests
* [x] websockets  
* [x] websockets tests

closes #1764
ref #1147

### CODE REVIEW

File: mitmproxy/proxy/protocol/http.py [hunk 1]
```diff
@@ -327,7 +330,15 @@ def _process_flow(self, f):
                 )
 
                 def get_response():
-                    self.send_request(f.request)
+                    if f.request.stream:
```
Comment (Kriechi) [inline mitmproxy/proxy/protocol/http.py]: Does `f.request.stream` get set to `True` anywhere? Or is this inherited from the flow?

Comment (ujjwal96) [inline mitmproxy/proxy/protocol/http.py]: Here:
https://github.com/mitmproxy/mitmproxy/blob/master/mitmproxy/addons/streambodies.py#L30

File: mitmproxy/proxy/protocol/http2.py [hunk 2]
```diff
@@ -487,9 +487,17 @@ def read_request_headers(self, flow):
 
     @detect_zombie_stream
     def read_request_body(self, request):
-        self.request_data_finished.wait()
```
Comment (Kriechi) [inline mitmproxy/proxy/protocol/http2.py]: We probably want to keep the wait for non-streaming requests (as you had before)

File: mitmproxy/proxy/protocol/http2.py [hunk 3]
```diff
@@ -487,11 +487,89 @@ def read_request_headers(self, flow):
...
+    def send_request_body(self, request, chunks):
+        if not self.no_body:
+            self.connections[self.server_conn].safe_send_body(
+                self.raise_zombie,
+                self.server_stream_id,
+                chunks
+            )
 
     @detect_zombie_stream
     def send_request(self, message):
```
Comment (Kriechi) [inline mitmproxy/proxy/protocol/http2.py]: `send_request(...)` now can just be:
```
  send_request_header(...)
  send_request_body(...)
```
Or did you change anything else?

File: test/mitmproxy/proxy/protocol/test_http1.py [hunk 4]
```diff
@@ -77,3 +79,35 @@ def test_head_content_length(self):
...
+            self.add_log(v)
+            return self.o.sendall(v)
+        else:
+            r = self.o.write(v)
+            self.add_log(v[:r])
+            return r
+
+
+class TestRequestStreaming(tservers.HTTPProxyTest):
+    @mock.patch("mitmproxy.net.tcp.Writer.write", side_effect=write, autospec=True)
```
Comment (Kriechi) [inline test/mitmproxy/proxy/protocol/test_http1.py]: If you move the mocking after creating the pathoc object, we should only affect new connections, but not pathoc...

File: test/mitmproxy/proxy/protocol/test_http1.py [hunk 5]
```diff
@@ -83,7 +82,7 @@ def test_head_content_length(self):
         assert resp.headers["Content-Length"] == "42"
 
 
-class TestStreaming(tservers.HTTPProxyTest):
+class TestRequest(tservers.HTTPProxyTest):
```
Comment (Kriechi) [inline test/mitmproxy/proxy/protocol/test_http1.py]: This still tests response streaming as well, why did you change the class name?

File: test/mitmproxy/proxy/protocol/test_http2.py [hunk 6]
```diff
@@ -862,3 +863,91 @@ def test_connection_terminated(self):
...
         assert connection_terminated_event.additional_data == b'foobar'
+
+
+class TestRequestStreaming(_Http2Test):
+
+    @classmethod
+    def handle_server_event(cls, event, h2_conn, rfile, wfile):
+        if isinstance(event, h2.events.ConnectionTerminated):
+            return False
+        elif isinstance(event, h2.events.RequestReceived):
```
Comment (Kriechi) [inline test/mitmproxy/proxy/protocol/test_http2.py]: I think there should be a `DataReceived` event which we should use instead.

File: test/mitmproxy/proxy/protocol/test_http2.py [hunk 7]
```diff
@@ -862,3 +863,91 @@ def test_connection_terminated(self):
...
+            req = rfile.read(109)
```
Comment (Kriechi) [inline test/mitmproxy/proxy/protocol/test_http2.py]: Why 109 bytes?

File: test/mitmproxy/proxy/protocol/test_http2.py [hunk 8]
```diff
@@ -862,3 +863,91 @@ def test_connection_terminated(self):
...
+            assert req
+            h2_conn.close_connection(error_code=5, last_stream_id=42, additional_data=req[9:])
+            wfile.write(h2_conn.data_to_send())
+            wfile.flush()
+
+        return True
+
+    def _send_request(self,
```
Comment (Kriechi) [inline test/mitmproxy/proxy/protocol/test_http2.py]: Please modify the existing `_send_request` function so we can change this with a new argument / config.

File: test/mitmproxy/proxy/protocol/test_http2.py (hunk 7 above)
Comment (ujjwal96) [inline test/mitmproxy/proxy/protocol/test_http2.py]: 9 bytes headers + 100 bytes bodies
Anything less would throw an error [here](https://github.com/mitmproxy/mitmproxy/pull/2369/files#diff-ae3ce6b83d0e48d6158f2e5fe6583e6bR911) inside `safe_read` because it then reads the length incorrectly

File: mitmproxy/addons/streambodies.py [hunk 9]
```diff
@@ -15,8 +15,15 @@ def configure(self, updated):
             except ValueError as e:
                 raise exceptions.OptionsError(e)
 
-    def run(self, f, is_request):
+    def run(self, f, is_request, is_wsflow):
         if self.max_size:
+            if is_wsflow:
```
Comment (Kriechi) [inline mitmproxy/addons/streambodies.py]: IMO in this case it is cleaner to use `isinstance(f, WebSocketFlow)` instead of a new bool argument.

**Comment by Kriechi** [comment]
@mitmproxy/devs this is ready for review!
IMO this is ready to merge, but I might have missed something important.

File: mitmproxy/proxy/protocol/http.py [hunk 10]
```diff
@@ -327,7 +330,15 @@ def _process_flow(self, f):
                 )
 
                 def get_response():
```
Comment (mhils) [inline mitmproxy/proxy/protocol/http.py]: `get_response` can be called twice (see [here](https://github.com/mitmproxy/mitmproxy/pull/2369/files#diff-1a48d22b11223fa300cac1e75cbcfafcR351) for the reasoning). If `stream` is set to True, the second send will have no body, because we have already depleted the body generator on the first attempt.

Unfortunately this is really hard to fix. I see two possible ways forward:

1. Store all chunks _somewhere_ so that we can play them again. That'd be nice as we then could always store the content while streaming, but that also adds substantial complexity.
2. Figure out a better way to see if the socket is closed. Maybe it's enough to just try reading from it with MSG_PEEK. Alternatively, maybe we can make `get_response` a `send_headers` function, and then do everything else (sending body + receiving response) outside of the try-except-repeat block. I am not sure if any of this works, we need to test that.

I think the second option is the generally nicer way (we can still implement 1 later), so we should probably try and improve that in a separate PR.

File: mitmproxy/proxy/protocol/websocket.py [hunk 11]
```diff
@@ -70,43 +71,50 @@ def _handle_data_frame(self, frame, source_conn, other_conn, is_server):
...
+                def get_chunk(payload):
+                    if len(payload) == length:
+                        # message has the same length, we can reuse the same sizes
+                        pos = 0
+                        for s in original_chunk_sizes:
+                            yield payload[pos:pos + s]
+                            pos += s
+                    else:
+                        # just re-chunk everything into 10kB frames
+                        chunk_size = 10240
```
Comment (mhils) [inline mitmproxy/proxy/protocol/websocket.py]: Minor suggestion: Make this a power of two, see e.g. the note in https://docs.python.org/3/library/socket.html#socket.socket.recv.

Comment (Kriechi) [inline mitmproxy/proxy/protocol/websocket.py]: Good point! @ujjwal96 please change it to 4096.

File: mitmproxy/proxy/protocol/http.py (hunk 10 above)
Comment (Kriechi) [inline mitmproxy/proxy/protocol/http.py]: Sounds all reasonable and we should fix this, but isn't this a general response streaming issue unrelated to this PR? If so, please open a new issue, and we can address it after merging this PR.

File: mitmproxy/proxy/protocol/websocket.py [hunk 12]
```diff
@@ -70,43 +71,50 @@ def _handle_data_frame(self, frame, source_conn, other_conn, is_server):
...
+                def get_chunk(payload):
+                    if len(payload) == length:
+                        # message has the same length, we can reuse the same sizes
+                        pos = 0
+                        for s in original_chunk_sizes:
+                            yield payload[pos:pos + s]
+                            pos += s
+                    else:
+                        # just re-chunk everything into 4kB frames
+                        chunk_size = 4096
```
Comment (Kriechi) [inline mitmproxy/proxy/protocol/websocket.py]: Thinking about it again, we should consider the frame header as well.
So the size of each chunk should be 4096 - header.
We also have to account for the direction a frame is sent, since the masking key is only for client->server frames.
https://tools.ietf.org/html/rfc6455#section-5.2

File: mitmproxy/proxy/protocol/http.py (hunk 10 above)
Comment (mhils) [inline mitmproxy/proxy/protocol/http.py]: As discussed on Slack, this is specific to this PR - `get_response` is retried, and on retry we have potentially already streamed some parts of the body.

File: mitmproxy/addons/streambodies.py [hunk 13]
```diff
@@ -28,12 +29,18 @@ def run(self, f, is_request):
...
 
-    # FIXME! Request streaming doesn't work at the moment.
     def requestheaders(self, f):
         self.run(f, True)
 
     def responseheaders(self, f):
         self.run(f, False)
+
+    def websocket_start(self, f):
+        if ctx.options.stream_websockets and isinstance(f, websocket.WebSocketFlow):
```
Comment (Kriechi) [inline mitmproxy/addons/streambodies.py]: I guess the `isinstance` check is not needed, who else would call `websocket_start`?

File: mitmproxy/addons/streambodies.py [hunk 14]
```diff
@@ -28,12 +29,18 @@ def run(self, f, is_request):
...
+            f.stream = True
+            ctx.log.info("Streaming WebSocket message {client} - {server}".format(
```
Comment (Kriechi) [inline mitmproxy/addons/streambodies.py]: From here on we are streaming all messages in this flow. The info text could be misleading.
Maybe rephrase to something like: **Streaming all WebSocket messages between {client} and {server}**

File: mitmproxy/options.py [hunk 15]
```diff
@@ -155,6 +155,13 @@ def __init__(self, **kwargs) -> None:
             """
         )
         self.add_option(
+            "stream_websockets", bool, False,
+            """
+            Stream WebSocket frames as and when they are received
+            by the proxy. If True message payload cannot be modified.
```
Comment (Kriechi) [inline mitmproxy/options.py]: Stream WebSocket **messages between client and server. Messages are captured and cannot be modified.**

File: docs/features/streaming.rst [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+As the name suggests, request streaming feature can be used to stream request body to the server incrementally
+before it has been fully received by the proxy. This may be useful for large file uploads.
+
+
+By default, mitmproxy will read the entire request/response, perform any indicated
+manipulations on it and then send the (possibly modified) request or response to
+the server or the client respectively. In some cases this is undesirable and you may wish to "stream"
+the request/response. When streaming is enabled, the request/response is
+not buffered on the proxy but directly sent to the server/client instead.
```
Comment (Kriechi) [inline docs/features/streaming.rst]: Move this paragraph to the top of the file, underneath a top-level headline: **HTTP Streaming**
Also mention that HTTP headers are still fully buffered before being sent.

Then the two sub-headlines for **Request Body Streaming** and **Response Body Streaming**.

File: test/mitmproxy/proxy/protocol/test_websocket.py [hunk 17]
```diff
@@ -328,3 +334,30 @@ def test_invalid_frame(self):
...
+
+        if not streaming:
+            with pytest.raises(exceptions.TcpDisconnect):  # Reader.safe_read get nothing as result
+                frame = websockets.Frame.from_file(self.client.rfile)
+        else:
+            frame = websockets.Frame.from_file(self.client.rfile)
+
+        if streaming:
+            assert frame
+            assert self.master.state.flows[1].messages == []
```
Comment (Kriechi) [inline test/mitmproxy/proxy/protocol/test_websocket.py]: We should asserting *something* if we are not streaming...
Or remove the `False` parameter option.

File: docs/features/streaming.rst (hunk 16 above)
Comment (Kriechi) [inline docs/features/streaming.rst]: For future reference: this was fixed with a new commit.

**Comment by Kriechi** [comment]
@mhils the HTTP/1 retry issue should be sufficiently resolved. Do we get a final seal of approval?
